File: Lab 11/11-3.java
import java.util.HashSet;
import java.util.Set;

class Solution {
    public int openLock(String[] deadends, String target) {
        Set<String> deadSet = new HashSet<>();
        for (String dead : deadends) {
            deadSet.add(dead);
        }

        if (deadSet.contains("0000")) {
            return -1;
        }

        Set<String> visited = new HashSet<>();

        int result = dfs("0000", target, deadSet, visited, 0);

        return result == Integer.MAX_VALUE ? -1 : result;
    }

    private int dfs(String current, String target, Set<String> deadSet, Set<String> visited, int steps) {
        if (deadSet.contains(current)) {
            return Integer.MAX_VALUE; // 
        }
        if (visited.contains(current)) {
            return Integer.MAX_VALUE; //
        }
        if (current.equals(target)) {
            return steps; // 
        }

        visited.add(current);

        int minSteps = Integer.MAX_VALUE;

        for (int i = 0; i < 4; i++) {
            char[] currentArray = current.toCharArray();

            currentArray[i] = currentArray[i] == '9' ? '0' : (char) (currentArray[i] + 1);
            String nextForward = new String(currentArray);
            minSteps = Math.min(minSteps, dfs(nextForward, target, deadSet, visited, steps + 1));

            currentArray[i] = currentArray[i] == '0' ? '9' : (char) (currentArray[i] - 2); 
                                                                                           
            String nextBackward = new String(currentArray);
            minSteps = Math.min(minSteps, dfs(nextBackward, target, deadSet, visited, steps + 1));
        }

        visited.remove(current);

        return minSteps;
    }
}

File: Lab 11/11-3.py
class Solution(object):
    def openLock(self, deadends, target):
        
        dead_set = set(deadends)
        visited = set()

        def dfs(position, steps):
            if position in dead_set or position in visited:
                return float('inf')  
            if position == target:
                return steps
            
            visited.add(position)
            
            min_steps = float('inf')
            for i in range(4):
                for direction in (-1, 1):
                    next_position = list(position)
                    next_position[i] = str((int(next_position[i]) + direction) % 10)
                    next_position = ''.join(next_position)
                    min_steps = min(min_steps, dfs(next_position, steps + 1))
            
            visited.remove(position)  
            return min_steps

        result = dfs("0000", 0)
        return result if result != float('inf') else -1
